core/storage/statistics/gae_models.py

```diff
@@ -1134,6 +1138,128 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+        return '%s.%s' % (exp_id, state_name)
+
+    @classmethod
+    def get_state_reference_for_question(cls, question_id):
+        """Generate the state_reference for the state in the question.
+
+        Args:
+            question_id: str. The id of the question.
+
+        Returns:
+            str. The state_reference for a new instance of this class.
+        """
+        return question_id
+
+    @classmethod
+    def create_model_instance(
+            cls, state_reference, entity_type, interaction_id,
+            learner_answer_info_list, schema_version,
+            accumulated_answer_info_json_size_bytes):
+        """Creates a new LearnerAnswerDetailsModel for the given entity type
+        then writes it to the datastore.
+
+        Args:
+            state_reference: str. The reference to the state for which model
+                instance is being created. For exploration state it will be of
+                the form 'exp_id.state_name', & for question it will be of the
+                form 'question_id'.
+            entity_type: str. The type of entity i.e "exploration" or
+                "question".
+            interaction_id: str.  The ID of the interaction for which the
+                answer details are received.
+            learner_answer_info_list: list. The list of LearnerAnswerInfo
+                objects in dict format, which is defined in the stats_domain.
+            schema_version: int. The version of LearnerAnswerInfo dict, which
+                is currently supported by the Oppia.
+            accumulated_answer_info_json_size_bytes: int. The size of the
+                learner_answer_info_list in bytes.
+        """
+        answer_details_instance = cls(
+            state_reference=state_reference,
+            entity_type=entity_type,
+            interaction_id=interaction_id,
+            learner_answer_info_list=learner_answer_info_list,
+            schema_version=schema_version,
+            accumulated_answer_info_json_size_bytes=(
+                accumulated_answer_info_json_size_bytes))
+        answer_details_instance.put()
+
+    @classmethod
+    def get_model_instance(
+            cls, entity_type, state_reference):
+        """Returns the model instance related to the entity type and
+        state reference.
+
+        Args:
+            entity_type: str. The type of the entity i.e "exploration" or
+                "question".
+            state_reference: str. The reference to a state, for which the model
+                is to be fetched. Foe exploration state it will be of the form
+                'exp_id.state_name', and for question state it will be of the
+                form 'question_id'.
+
+        Returns:
+            LearnerAnswerDetailsModel. The answer details model associated
+                with the given entity type and state reference. Doesn't include
+                deleted entries.
+        """
+        model_instance_list = cls.get_all().filter(
```

Comment:
Ideally this is a get-by-ID operation. You should be able to do that if you actually construct the model IDs deterministically, so that you can reconstruct them here when needed and retrieve the correct model.

Reply:
Okay, I was looking to how to generate the ID of the model. As we want to reconstruct it again so we can't random things. I thought of to use ```entity_type.state_reference```, but it would fail as:
Case1:
exp_id = 'exp.id'
state_name = 'state'
id = 'exploration.exp.id.state'

Case2:
exp_id = 'exp'
state_name = 'id.state'
id = 'exploration.exp.id.state'

So in this case it will fail, any suggestions?

Reply:
Got the idea, I will use the hashlib library.

Reply:
But yeah for hash I need to specifically pass the exp_id and state_name (question_id for question) which will not be feasible.
What do you suggest @seanlip and @aks681 ?

Reply:
I don't think the cases you told above would occur, as exp id itself is a random hash. So, it might be fine to just use exp id and state names.

Reply:
Oh, then it's fine. Thanks :)

Reply:
Done

Reply:
This is good thinking, and you should use a delimiter that is not used in the charset of either the exp ID or the state names when constructing the composite ID, so that the names can be parsed uniquely. You should not need to encode either of these components though.

Reply:
I think storing can be done with ```.``` delimiter as exp_id has its own hash value so the model id will be unique. Okay I will test it for unicode state name and check if it can find the model instance.